Tests/WeaverDITests/ThreadSafeInstanceCachingDecoratorTests.swift

```diff
@@ -0,0 +1,80 @@
+//
+//  ThreadSafeInstanceCachingDecoratorTests.swift
+//  WeaverDITests
+//
+//  Created by Bruno Mazzo on 5/31/18.
+//
+
+import Foundation
+
+import XCTest
+
+@testable import WeaverDI
+
+class ThreadSafeInstanceCachingDecoratorTests: XCTestCase {
+    
+    
+    let instanceKey = InstanceKey(for: String.self, name: "test")
+    
+    func builder() -> String {
+        return "builder"
+    }
+    
+    func test_set_should_delegate_to_inner_builder() {
+        let cacheMock = InstanceCacheMock()
+        
+        let sut = ThreadSafeInstanceCacheDecorator(instances: cacheMock)
+        
+        _ = sut.cache(for: instanceKey, scope: .container, builder: self.builder)
+        
+        XCTAssertEqual(cacheMock.cacheCallCount, 1)
+        XCTAssertEqual(cacheMock.scope, .container)
+        XCTAssertEqual(cacheMock.key, instanceKey)
+    }
+    
+    func test_many_access_in_parallel() {
```

Comment:
Let's write this test the same way it's written in `DependencyContainerTests.swift`